### ISSUE TITLE
Safari (WebKit) Support

### ISSUE THREAD

**Issue by amirrustam** [cypress-io/cypress#6422]
> **Current Status** - _last updated August 4, 2022_
> 
> ✅ Prioritized and on the roadmap
> ✅ Doing exploratory work to assess requirements and timelines
> ✅ Under active development
> ℹ️ _Cypress currently supports [Firefox](https://cypress.io/blog/2020/02/06/introducing-firefox-and-edge-support-in-cypress-4-0/) and Chrome-family browsers (including Edge and Electron)_.

**Objective**
---
Add support for running Cypress tests within WebKit, which is the engine used by the [Safari](https://www.apple.com/safari/) web browser.

**Task List**
---

- [x] Initial PR to add development-only support: https://github.com/cypress-io/cypress/pull/15533
- [ ] Fix required driver/other functionality
	- [x] Network functionality https://github.com/cypress-io/cypress/pull/23232
	- [x] Keyboard tests (`type*.cy.js`) https://github.com/cypress-io/cypress/pull/23522
	- [x] Input tests (`click.cy.js`, focus tests...) https://github.com/cypress-io/cypress/pull/23522
	- [x] Stack traces https://github.com/cypress-io/cypress/pull/23730
	- [x] Check component testing
	- [x] Video recording #23579 
	- [x] Handle downloads https://github.com/cypress-io/cypress/pull/23662
	- [x] Fix global mode detection of npm module https://github.com/cypress-io/cypress/pull/23711
	- [ ] "Network" GraphQL error fix - wip
	- [x] [Fix open mode close](https://github.com/cypress-io/cypress/pull/15533#issuecomment-1211506734) https://github.com/cypress-io/cypress/pull/23662
	- [x] Disable `forceNetworkError` https://github.com/cypress-io/cypress/pull/23711
	- [x] Fix cross-origin unhandledRejection events: https://github.com/cypress-io/cypress/pull/23588
- [ ] Possibly fix less critical functionality
	- [ ] Multi-domain implementation https://github.com/cypress-io/cypress/pull/23442
	- [ ] Fix websocket polling/proxying/allowedSockets
	- [ ] [Fix devicePixelRatio](https://github.com/cypress-io/cypress/pull/15533#issuecomment-1210169765)
	- [x] Fix/skip focus from app
- [x] Documentation - limitations, Linux dependencies https://github.com/cypress-io/cypress-documentation/pull/4710
- [x] Add experimental flag, make available in production, update public types https://github.com/cypress-io/cypress/pull/23711

**Comment by foolip** [comment]
I've done a fair bit of work on getting Safari working well with [web-platform-tests](https://github.com/web-platform-tests/wpt) (a conformance test suite for browsers with a [dashboard](https://wpt.fyi/)) and know a little bit about what kinds of problems you might run into.

I know that Cypress doesn't use WebDriver but rather CDP (at least for Chrome) but can someone sketch roughly what the requirements of the protocol are? I've had a look at CONTRIBUTING.md and [browser support docs](https://docs.cypress.io/guides/guides/launching-browsers.html) but couldn't find any high-level information about Cypress internals.

I ask because I'd like to understand if it would be at all feasible to communicate with Safari using safaridriver, or what the basic "API" or protocol between Cypress and a browser needs to be.

**Comment by Bkucera** [comment]
hi @foolip , we'd need the interfaces found here https://github.com/cypress-io/cypress/blob/6790154557e7c85d5d90ef9ce202bcfaca6e78bd/packages/server/lib/browsers/cdp_automation.ts#L129-L181, as well as a method of recording video.

so a brief list would be:
- get/set/clear cookies by domain
- screenshot page
- recording video (this can be done via getUserMedia api if it supports capturing the browser tab, which we do in firefox)

**Comment by foolip** [comment]
Thanks @Bkucera, that's a great list!

> get/set/clear cookies by domain

It sounds like you need to get/set/clear cookies for domains other than the one currently loaded, is that right? It looks like the only way of doing that using WebDriver/safaridriver would be to load a page from that domain and then use the [endpoints](https://w3c.github.io/webdriver/#get-all-cookies) with an implicit domain, but that's not great.

> record video of page

I don't think this is possible, but I did have [fun](https://github.com/web-platform-tests/wpt/pull/20007) recording the whole macOS VM in Azure Pipelines using `screencapture` once. One can also record a specific window, but I presume you want the content area without decorations.

> a way to set a proxy server for all network traffic

I'm pretty sure this is not possible yet.

> a way to bypass the invalid ssl cert screen/prompt

I wouldn't be surprised if this is already possible with some setting or command line argument, but I don't know what it is. Our setup for web-platform-tests is a bit special, we update `/etc/hosts` and add a certificate to the system certificate store. It's a bit much, and also requires that you know what the domains under test are before you start the tests, which is probably not the case for Cypress.

**Comment by kaushalyap** [comment]
@amirrustam We got Firefox and Edge support recently, love to see, Safari support soon since Safari have

- 17.21% global browser market share (2nd most used browser, second only to Chrome) : [source](https://gs.statcounter.com/browser-market-share)
- 36.28% USA browser market share (2nd most used browser in USA, second only to Chrome) : [source](https://gs.statcounter.com/browser-market-share/all/united-states-of-america)
- 33.48% UK browser market share (2nd most used browser in UK, second only to Chrome) : [source](https://gs.statcounter.com/browser-market-share/all/united-kingdom)

So please support Safari too.

**Comment by kelly-tock** [comment]
Yes for a lot of responsive sites mobile safari is a must have and can be the majority of the users.

**Comment by clockelliptic** [comment]
i want so badly to use Cypress, but I simply must use WebDriver because Safari support is a requirement over here

**Comment by Lughino** [comment]
I’d love to see a quick decision around this topic, especially now that playwright is gaining a lot of popularity.

Cypress team, please prioritise this compatibility.

**Comment by it-sha** [comment]
I would suggest to my company and all clients to use Cypress if Safari support was implemented

**Comment by lkoorsse-tails** [comment]
Starting using cypress instead of selenium, great advantages over selenium, however Safari browser is a major requirement, would love to see cypress support for this soon to continue using cypress

**Comment by AlexanderTsipin** [comment]
same, using Wdio just because safari support, management requirements.
Please implement even a basic safari runner.

**Comment by flenerob** [comment]
Please add support to Safari !!!

**Comment by jhuesos** [comment]
Could this issue to be locked for comments. Adding the "I really need this message" does not really help and it creates a new push notification that everyone subscribed receives and it is annoying. Do you want to reflect that you need this? Click 👍  on the issue.

**Comment by mydiemho** [comment]
In the meantime, could we update the error to show that Safari is not supported in Cypress rather than the current misleading error that it isn't installed.  I wasted a few hours trying to debug my CI because I thought Safari was not installed. 

Here's the current error message

```
> npx cypress run --browser  "safari"

Can't run because you've entered an invalid browser name.

Browser: 'safari' was not found on your system.

Available browsers found are: chrome, firefox, electron
```

**Comment by estefafdez** [comment]
I would love to have that!

**Comment by AlexanderTsipin** [comment]
Found Solution: TestCafe supports Safari and iphone, real devices and emulators

**Comment by wojciech-sygier** [comment]
I'm gonna add +1 to people requiring Safari support in Cypress. We're currently experimenting with different frameworks and we narrowed down our selection to TestCafe and Cypress. While most of us wants Cypress, we also need the ability to test on Safari (due to ~30% of our users using it, 77% of which are using mobile version).

Can we have at least some rough estimate for when any work would start? SoonTM is too broad and we can't really risk rewriting our tests if we don't get it in, let's say, a year or so.

**Comment by 0xc0d3r** [comment]
@wojciech-sygier We too have the similar user segment and we took a call to proceed with TestCafe and it is working for us. We are using BrowserStack to run the tests on real devices and it is super cool.

**Comment by galeza** [comment]
Please, 
Add suport for Safari!!!

**Comment by jennifer-shehane** [comment]
We evaluate current feature work on a number of factors and have to make tough decisions on which to currently work on. This is not part of our current feature work. This is not to say that it will never be delivered. We will comment in this issue when that decision changes. We do not have a rough estimate on when this will be delivered.

We are always re-evaluating what to work on next and feedback from people using Cypress is a big part of that evaluation. We welcome comments in this thread on how this feature would impact your testing.

**Comment by jennifer-shehane** [comment]
@mydiemho This error message was updated to be clearer about which browsers are supported in Cypress 4.8.0. https://github.com/cypress-io/cypress/pull/7375

**Comment by kelly-tock** [comment]
Where can we see the current feature work that is prioritized?

**Comment by aikhelis** [comment]
+1 for Safari su9pport has been long awaited, as second most used browser (Mac and iOS/iPadOs) for us. Getting to a point of considering moving all the existing test coverage in cypress to testcafe for browser support.

**Comment by jennifer-shehane** [comment]
Our [Roadmap](https://on.cypress.io/roadmap) highlights some of the larger features that are under development.

**Comment by aikhelis** [comment]
Thank you @jennifer-shehane - this is useful and I'm sure some of these features are important, as long as users are not switched to another tool for the coverage of their user real experience across key browsers.

**Comment by apis3445** [comment]
+1 for safari support browser

**Comment by jprealini** [comment]
+1 for Safari support

**Comment by hitmands** [comment]
Hi @jennifer-shehane, 

I think this is more around showing open commitment on providing support for safari, 
it's quite concerning that the official roadmap doesn't mention it at all and this topic still says:

![image](https://user-images.githubusercontent.com/5674833/89050416-a5261b00-d34a-11ea-80c1-c4c900d3484f.png)

___

It would be great to have this moving forward, 
at least in terms of direction.

**Comment by sharmilajesupaul** [comment]
I'm curious if the team is also considered adding standalone [WebKit](https://webkit.org/) support in addition to Safari for MacOS?

One of the big benefits is that WebKit is Linux compatible. Which means CI systems or Docker environments can make use of this automation. It's also the browser backing most mobile browsers on iOS which is an important part of coverage for most companies.

We've been experimenting with Playwright as well and they seem to support WebKit out of the box https://github.com/microsoft/playwright#documentation--api-reference.

**Comment by mgol** [comment]
> I'm curious if the team is also considering adding standalone [WebKit](https://webkit.org/) support in addition to Safari for MacOS? (I can't seem to find any issues about this).

@sharmilajesupaul If nothing changed since I last checked, WebKit is way less uniform than Blink. Browsers using Blink generally depend on Chromium without major changes to the engine so they all render web pages identically. WebKit, on the other hand, has multiple build flags so various WebKit-based browsers may be quite different between each other. That’s one of the reasons Safari & Chrome have always had many differences in supported Web APIs.

Maybe it changed since Blink was forked out of WebKit, I don’t know.

**Comment by sharmilajesupaul** [comment]
@mgol Thats interesting, I hadn’t considered Blink since chrome browsers are well supported by Cypress. My thinking was around the fact that WebKit with an emulated device would probably be the closest thing to mobile Safari that we could run in  Docker environments.

**Comment by hehoo** [comment]
@amirrustam Could you share the current status for this issue? Have we had a plan for implementing the support for Safari? Recently we got a request from our product team asking us to run our E2E tests on Safari. It will be great if we got it planned 😄 .

**Comment by Ekaterina-Krovyakova** [comment]
@amirrustam Hi ! I also agree that many QA engineers are waiting for Safari browser support. Then Cypress could be the tool of choice for many teams.

**Comment by WilliamForbesJr** [comment]
Love the world you all are doing in Cypress! I would certainly like to add my input here as well! We encountered some Safari-specific bugs recently that would have been absolutely caught had we had Cypress helping to cover it. Safari would be a game-changer!

**Comment by sovkan** [comment]
Not sure if this is the place, but I wanted to throw my hat in as well with the need for Safari support. The company I work for ended up not going with Cypress due to no safari support.

Big sad.

**Comment by codinguncut** [comment]
> 
> 
> hi @foolip
> we would need the following from either extension/devtools APIs:
> 
>     * get/set/clear cookies by domain
> 
>     * screenshot page
> 
>     * record video of page (this can be done via getUserMedia api if it supports capturing the browser tab, which we do in firefox)
> 
> 
> in addition, we would also need:
> 
>     * a way to set a proxy server for all network traffic
> 
>     * a way to bypass the invalid ssl cert screen/prompt
> 
> 
> everything else is done via JavaScript

Is there any chance of having partial/ experimental Safari support?
- without get/ set/ clear cookies
- using a valid cert authority (+/- HSTS) or only against HTTP
- using a pre-set/ manually-set proxy

**Comment by kennyg1989** [comment]
Safari is one of the most widely used browsers in the market, this should be prioritized!

**Comment by codinguncut** [comment]
@Bkucera I would be happy to work on this.
Could you point me to a good starting place in the code base (i.e. different implementations of this for Chrome, FF, Edge)?
Do you know if anyone has already started work, or achieved partial results for this?

**Comment by kristianmandrup** [comment]
Check out:
- https://webkit.org/blog/6900/webdriver-support-in-safari-10/
- https://www.w3.org/TR/webdriver/

**Comment by ayang3** [comment]
We just migrated to Cypress, but we are still using Selenium for Safari testing. Given the high volume of mobile web users on Safari, supporting testing in Safari with Cypress would be a huge win for us, and we can completed migrate to Cypress with cross browser testing! Would love to see the support of Safari in Cypress.

**Comment by danawoodman** [comment]
For those looking for a solution to Safari support, I highly recommend you checkout [Playwright](https://playwright.dev/). Seems like a good alternative for those of us that aren't ok with not testing Safari 👍

**Comment by krysalead** [comment]
Hi,

We have internally a discussion about Playwright vs Cypress, I am fighting for Cypress which I found syntax quite nicer than Playwright (too much coding style). But today they have an argument regarding the support of Safari. So, it become political :).
Good luck for the implementation

**Comment by VickyLund** [comment]
We started with Cypress 6 months back and we now have a suite of 750+ cases.
We face a lot of issues with Safari, especially in Webviews.
A good percentage of our users use Safari to browse our website.
Would love to see this being prioritized as we'll be able to use existing cases to cover all major flows in Safari immediately.

**Comment by Olyya** [comment]
HellO!
Cypress is an amazing tool. But for suggestions using cypress for customers need support safari. Is there any estimation about this, or do you plan to do that at all?

**Comment by jkavalec** [comment]
Adding Firefox and Edge was amazing thing! And for other amazing things like Component testing, really good job on that too.

However as all Apple things Safari is horrible thing that just must do things differently than everyone else, so support for Safari is many times more important than Firefox and Edge, in those two you can generally expect it will work similar to Chrome, but Safari that is quite different story.

I really hope we can convince you that Safari support is part of Cypress that is most missed.

**Comment by DanielBoa** [comment]
I'd love to be using Cypress but it's a hard sell without it Safari support.

Look forward to the day it's announced!

**Comment by cbrunnkvist** [comment]
I thin the biggest motivation for figuring this one out would be that **_once we support Safari, then support for mobile Safari in Simulator/wired device should not be too far off either!_** 💯 🥇 🍍

**Comment by gutoslv** [comment]
Unfortunately, we need to cover Safari within our tests since we have a significant portion of users who have Safari as the main browser. I'd love to use Cypress as the main automation tool at the company I work for, but it's hard to convince the team without Safari support. Would be easier to convince them if we had it mapped at least on the roadmap.

The work you guys are doing is amazing, and Cypress is definitely the best automation framework that I've worked with. Hope you guys support Safari soon!

**Comment by hogarthww-labs** [comment]
I've started development on a [Cypress DSL adapter for Playwright](https://github.com/hogarthww-labs/cypress-playwright-api-adapter). Feel free to contribute.

**Comment by dylanz666** [comment]
Hope cypress can support Safari !

**Comment by vitaliifokine** [comment]
It's almost 2021. And no support for Safari. Seems Cypress is at a dead end. It's a critical feature.

**Comment by wojciech-sygier** [comment]
I have two requests regarding this issue (from the user's perspective):
- (for Cypress team) Could you update this feature's status as of now? Last update is from the beginning of this year.
- (for users) As some people said before me - add more thumbs up for this issue so the team sees how important it is. Also, provide arguments as why you need this to be included. Just saying "please add it" or mocking the team or the product won't necessarily take us anywhere.

I'll update the needs of my team - despite no support for Safari we decided to go with Cypress anyway, we keep adding more tests in place of our old framework (we're at around 200 cases now). However, we do have a lot of users on Safari, who are also important for us. Despite very little Safari-specific bugs showing up recently we need to cover that browser too. In the long run lack of Safari support may eventually become somewhat problematic.

**Comment by cbrunnkvist** [comment]
I have a feeling that only once Safari-support feature requests from actual
*paying* Cypress Dashboard users become high enough, priority will
increased ;-)

Den tis 24 nov. 2020 kl 16:02 skrev Wojciech Sygier <
notifications@github.com>:

**Comment by lvkron7809** [comment]
I truly believe Cypress is the Gold Standard in Web-UI testing. However, I can't get the head of our QA to stick with Cypress, even though everyone knows it's better than everything else out there by a long shot, because it doesn't support Safari. This has been an ongoing debate in our QA org.  I have been viciously fighting to keep Cypress but without Safari I can't seem to convince the leadership to keep it based solely on that fact. Instead we are being forced to implement some hand-rolled Python/Selenium junk that I really want no part of. Not because of Python, but 100% because I think Selenium is junk and is a pain to work with. I have nothing really but negative experiences with Webdriver or Webdriver.io, Selenium etc..

Please for the love of all things holy, make support for Safari a thing. 

You guys are great, keep up the awesome work. Love me some Cypress.

**Comment by jennifer-shehane** [comment]
Safari support is still under consideration and we are gathering interest from this thread to help us weigh this feature against others. We have hard decisions to make about what to work on and have multiple ways that we gather feedback, one of which is this issue thread. So we're really thankful for everyone that has expressed interest. 🙏

Priorities can change as we move forward, but I've updated our Roadmap to show the larger features that we plan to work on in the next few months: https://on.cypress.io/roadmap

**Comment by ericpoulinnz** [comment]
Keen to know how other features on the roadmap come in ahead despite this feature clearly having much more interest?  How important a factor is interest from your users in deciding priorities?

Safari is the second most used browser globally and has many differences to Chrome or Edge which we absolutely need our end-to-end tests to cover.

**Comment by loicortola** [comment]
Hello,
just adding a +1 on this thread.
Although Safari is not the #1 browser, supporting it is a must-have for lots of organizations (sometimes really just for political reasons). My company is a service company and Cypress is not used anywhere in their clients, forced to still used Webdriver based solutions.
@jennifer-shehane , I saw the roadmap which looks very interesting in terms of features. I would probably move it above iframe support ;)

**Comment by SleepWalker** [comment]
First of all I would like to thank @loicortola, @ericpoulinnz, @lvkron7809  and friends for their valuable comments. It is really important for us (hundreds of people following this issue) to know that your pain is the strongest. Thank you very much!

@amirrustam it will be cool if you lock the discussion. This spam makes no sense and only distracts everyone. It's also really hard to find really important messages in this thread.

**Comment by rakeshsukla53** [comment]
Please add support for Safari. We love Cypress but we continue to encounter bugs on Safari which can't be reproduced on Chrome or Firefox.

**Comment by jennifer-shehane** [comment]
@SleepWalker You can unsubscribe from this thread to stop receiving notifications on the righthand side of the issue. Our team wants to gathering feedback from the community and will not be locking the issue to conversation. 

<img width="339" alt="Screen Shot 2020-12-07 at 10 36 16 AM" src="https://user-images.githubusercontent.com/1271364/101308375-1ab17300-3878-11eb-8141-955ae5e30b83.png">

**Comment by rsbabcock** [comment]
While additional support for safari would be ideal. A potential solution is to offer a more in depth tutorial on how to use .task() to run a safari webdriver. The line about spinning up a safari webdriver makes me think I am able to test safari using cypress, by using a plugin. Adding an example would allow developers who love cypress but must have safari support a viable option to use Cypress.

https://docs.cypress.io/guides/tooling/plugins-guide.html#cy-task
![Screen Shot 2020-12-09 at 1 59 25 PM](https://user-images.githubusercontent.com/36278605/101680852-e92ada00-3a26-11eb-8756-8fc0c087380f.png)

**Comment by alejandroiglesias** [comment]
@rsbabcock this could be very interesting, although I assume it wouldn't be trivial. You would need to deal with the Webdriver protocol and do a lot of boilerplate by yourself. I might be wrong, would like to know what are the possibilities of `cy.task` as well.

**Comment by maga-polito** [comment]
Many of our clients use Safari and right now our automated tests cannot run over safari, this feature should be considered

**Comment by zslucky** [comment]
We faced same issue, need safari a lot 👍

**Comment by hsiaoa** [comment]
> I have a feeling that only once Safari-support feature requests from actual *paying* Cypress Dashboard users become high > enough, priority will increased ;-) 

I am adding my +1 here, and no joke, if there's a "Pay for Safari Testing" button our team will gladly click it. Please add it on the roadmap.

**Comment by greenfish-1** [comment]
> > I have a feeling that only once Safari-support feature requests from actual _paying_ Cypress Dashboard users become high > enough, priority will increased ;-)
> 
> I am adding my +1 here, and no joke, if there's a "Pay for Safari Testing" button our team will gladly click it. Please add it on the roadmap.

My +1 too. We really need Safari support. Being able to use `cy.task` to launch other browsers may be better than nothing but can't compare it with having Safari fully supported. Other teams in my organisation want to use Selenium because it supports Safari and IE11 and if Safari support gets added to the roadmap, it'll be a strong argument for using Cypress.

**Comment by whoiskatrin** [comment]
Not having a Safari support leaves us again with Selenium, which is not even half as good as Cypress. It makes it very hard to switch to and convince the whole organization to do so.

**Comment by jennifer-shehane** [comment]
We want to thank everyone for your comments. 🙏 They are helpful for our team in understanding what features are important for us to investigate. We encourage anyone coming across this to continue to share their use case as well.

We wanted to explain some of the technical difficulties of building support for the Safari browser that are uniquely challenging for Cypress.

The main challenge being Safari's implementation of ["Glass Panes"](https://developer.apple.com/documentation/webkit/about_webdriver_for_safari)

>To prevent any attempts to interact with the window or web content during a test, Safari installs a transparent “glass pane” over the automation windows while the browser is being used for WebDriver testing. This pane catches any stray interactions (mouse, keyboard, resizing, and so on) that could affect the automation window. https://developer.apple.com/documentation/webkit/about_webdriver_for_safari

One of Cypress's main features is the ability to run your tests in an 'interactive' mode (via `cypress open`). It allows you to interact directly with your application under test to help write and debug tests in real time. Cypress does not just run in a headless environment. The Glass Panes would make this interactive experience impossible in Safari directly.

Cypress is not one to shy away from technical challenges that seem 'impossible' though, so we are continuing to investigate options to provide full cross-browser testing. There are a few directions that we've been discussing to move forward with a 'Safari-like' testing experience. 

**A noninclusive list of directions we could take:**

- Only support Safari during `cypress run`. This does not really live up to our standards of the experience we want to offer with Cypress. It will degrade the true TDE/debugging experience of Cypress when testing in Safari. It is being considered though.
- Support a 'Webkit' browser. Webkit is the rendering engine that powers Safari. We could offer testing within a Webkit-powered browser. And extend it to be close to the same as testing in Safari.

We’d love to hear any thoughts on whether either of the above options would satisfy your needs for Safari testing.

**Comment by krysalead** [comment]
Hi, thanks for the explanation and from my point of view the runtime is more important, we can still open the browser and do a manual scenario to reproduce and debug. So first solution having this safari support only for cypress run is quite sufficient. Thanks again for your hard work.

**Comment by hsiaoa** [comment]
> Hi, thanks for the explanation and from my point of view the runtime is more important, we can still open the browser and do a manual scenario to reproduce and debug. So first solution having this safari support only for cypress run is quite sufficient. Thanks again for your hard work.

Totally second this. TBH most of the time I dev and debug on chrome, and it's crucial for cypress run to catch safari specific bugs so I can open safari to see what's going on.

**Comment by dverbiest** [comment]
Fully agree, run-only better than a 'fake' Safari. That could come after it
to debug Safari a bit better?

Op di 22 dec. 2020 07:34 schreef Krys Alead <notifications@github.com>:

**Comment by VickyLund** [comment]
For us, having true Safari only via `cypress run` would still add 85% of the value against having entire Safari experience.
But being able to test in Safari is critical for us so we would be fine with the first option.
Though, I guess, we'll still have all failure images and videos available for debugging/simulating scenarios manually, right?

**Comment by mrmckeb** [comment]
@whoiskatrin, in the meantime you can look to Microsoft's [Playwright](https://playwright.dev/) and [Playwright test runner](https://github.com/microsoft/playwright-test/tree/master/examples) instead of Selenium. It has many of the same features as Cypress, but without the UI and a lower-level DX.

@jennifer-shehane have you reached out to the Playwright team? Maybe you could collaborate with them, as they're maintaining a fork of Webkit.

I feel that supporting Webkit, as Playwright does, would be significantly better than nothing.

**Comment by walterbrand** [comment]
Thanks Jennifer for clarifying the problems you run into when integrating Safari support. From my end I would argue that catching regression bugs in Safari is much more important than live debugging. Not having the ability to test in Safari is a blocker to switch. Not being able to debug is a nice to have. So only being able to run cypress run would suffice.

**Comment by gutoslv** [comment]
For us the Safari run-only would be better than a "fake" Safari experience.

**Comment by literalpie** [comment]
For our uses, running the tests is most important and the live interaction isn't needed.

Also, any testing in WebKit would make me pretty confident.

Either of the two compromises would be good enough and much better than the current state. Given the choice, I think the first approach would preferable.

**Comment by sovkan** [comment]
> * Only support Safari during `cypress run`. This does not really live up to our standards of the experience we want to offer with Cypress. It will degrade the true TDE/debugging experience of Cypress when testing in Safari. It is being considered though.
> * Support a 'Webkit' browser. Webkit is the rendering engine that powers Safari. We could offer testing within a Webkit-powered browser. And extend it to be close to the same as testing in Safari.


The first option would be best in my opinion. Getting Safari support at all is tremendously better than not having it at all. No Safari is the sole reason the company I am with chose to go with Nightwatch instead. We do not need the visual debugging aid for Safari if that is the primary reason for not doing it. Solve that issue another day. We can rely on screenshots/other debugging means if we see Safari tests failing. It would be no different than any other framework, as far as Safari is concerned.

**Comment by marktnoonan** [comment]
One of the main use cases for the visual interaction is that you can involve cypress in your development process in a more intimate way than just as a test-runner. I do think it's OK if Safari is not supported for this use-case, as probably there are not that many front-end developers really relying on Safari as their daily driver for web development, vs Chrome and Firefox. It's also not so rare to mix and match browsers for the different capabilities - eg using Firefox to debug CSS Grid issues and Chrome for other stuff. So developers who prefer Safari aren't exactly denied the experience.

I don't see the two options as exclusive - eg, real Safari could be supported during `cypress run` and a webkit alternative could be made available for interactive debugging. And hopefully one day Apple allows test runners to remove the glass pane and Safari can be supported everywhere.

In a lot of places, I'm betting reasons like "the execs visit the product on iPhone" drives some reluctance to omit Safari from testing since Safari bugs might actually be the most visible to certain people in the org.

**Comment by lyjo23** [comment]
Thank you Jennifer for the explanation and taking this onboard.
For us, being able to run in real Safari would be more beneficial than running in Webkit browser. Being able to run tests in Webkit browser is still okay(not sure what the exact difference between Webkit browser and Safari would be though?) but ultimately we'd like to be able to run our e2e tests in Safari to have accurate representation of the user experience.

**Comment by lvkron7809** [comment]
I put a thumbs up on @marktnoonan response but I thought I'd throw my 2 cents in in agreement with his comment. Most notably this, "real Safari could be supported during cypress run and a webkit alternative could be made available for interactive debugging". While Cypress' interactive debugging is second to none and extremely helpful and important, at least being able to run against Safari and debug in other ways would be a big stop gap.

**Comment by rakeshsukla53** [comment]
👍  for the first option

**Comment by thecb4** [comment]
+1 for Safari/WebKit with cypress run.

**Comment by rsbabcock** [comment]
+1 for Safari/Webkit with cypress run. Our team's use would be purely to satisfy product requirements that we "support" safari and ensure that everything runs fine. We would basically only use safari results to catch safari errors, then fix them locally so not really a huge deal to have that debugging in real time with Cypress.

**Comment by jacobprouse** [comment]
+1 for Safari/WebKit with cypress run, our team is in the same position as @rsbabcock

**Comment by tstecca** [comment]
Hello Cypress.  Thank you for this excellent testing library.  Our company recently migrated to Cypress from Nightwatch to test a few of our products and we are very happy with the Cypress.  I note from the comments above that you are using this thread to gauge interest in this feature.  The lack of Safari support is an issue for us.  Safari represents a large proportion of our user base and also a large proportion of edge cases that we deal with.  Safari support would be our preferred new feature for Cypress.  We find it rather feature-complete otherwise.  Partial support would be fine. We would be satisfied with just having the tests run against Safari/webkit.  Other features like video capture and interactive debugging could come later.  Thanks again for Cypress.

**Comment by EVaisman** [comment]
+1 for Safari support

**Comment by r-moore** [comment]
+1 for Webkit / Safari

**Comment by felippepuhle** [comment]
+1 for Safari

**Comment by aaujayasena** [comment]
+1 for Safari/IE

**Comment by 8thblock** [comment]
Safari has a huge chunk of the market. Omitting it from your lib is show stopper for many shops. In fact it is silly to use Cypress until Safari is fully supported

**Comment by VickyLund** [comment]
Is making Safari available with Cypress for both mobile and desktop one unit or would it be two different tasks?
If it's the latter, we would prefer to have mobile support first.

**Comment by 8thblock** [comment]
> Is making Safari available with Cypress for both mobile and desktop one unit or would it be two different tasks?
> If it's the latter, we would prefer to have mobile support first.

Mobile browsers is outside the scope for these e2e tools. That's because it would mean: starting the emulator, opening a browser there, navigating the flow etc. If you want full e2e tests on real mobiles the likes of BrowserStack or Sauce lab is you friend

**Comment by cm0s** [comment]
Here in Switzerland (in our Universities) Safari is used by 35-40% of our users. Being able to test Safari even only with cypress run will be extremely valuable.

**Comment by daiky00** [comment]
@amirrustam I think this a very good featured cypress should support to be one of the top e2e testing frameworks please consider it. It has been a year and not movement on the topic from cypress development team.

**Comment by amirrustam** [comment]
@daiky00 we appreciate your feedback and be assured that we hear you. The team is currently/actively investigating this feature. So there is movement on this internally, even if there is no GH activity at the moment.

**Comment by pete-om** [comment]
We have safari-only rendering issues that occasionally affect our product, so safari support in Cypress would be a real boon. Either a webkit alternative or safari via cypress run only (assuming that includes video recording) would both be suitable for us.  Cheers!

**Comment by RajaBellebon** [comment]
+1 for supporting Safari, in our different projects, we are still using a mix of WebdriverIO + Browserstack to run cross browser tests, especially for Safari and Edge (We don't work with windows machine but our consumer do). We do have a bit of Cypress coming into the mix but definitely interested by this feature as we could move away from different tech stack and tools and only use Cypress.

**Comment by paolarosanarodrigues** [comment]
@jennifer-shehane any news about this topic?

**Comment by jamesbegg** [comment]
To echo (but reaffirm) many others, I absolutely want to use Cypress and have enjoyed using it vastly more than any other testing tool, however this lack of webkit/safari support is not an acceptable barrier to my project. You would *almost* monopolise the E2E testing market if you can release this.

**Comment by becciwatson** [comment]
> 
> 
> To echo (but reaffirm) many others, I absolutely want to use Cypress and have enjoyed using it vastly more than any other testing tool, however this lack of webkit/safari support is not an acceptable barrier to my project. You would _almost_ monopolise the E2E testing market if you can release this.

Hi James, I agree, this is really such a fundamental feature for us too, that we have had to get an additional suite.  And it has made me review switching vendors.  Cypress doesn't seem to be in touch with customer needs and frankly, we're sick of waiting.

**Comment by linuxkidvoid** [comment]
Given mobile webs are so important, and safari ipad combination will go on for a while for most of our customers, it's critical we support Safari. This is a trend will go on for a while.

**Comment by r-moore** [comment]
Some numbers that emphasise the importance of this... in the USA, Safari usage on mobile platforms is something like 52% vs. 40% for Chrome. 
In the UK it's ~48% Safari, ~41% Chrome. 
Where I live in Jersey, Safari is 67%, Chrome 25%.
[Source: statcounter.com]

It's obviously super important to be able to test the dominant browser - or at least the javascript/rendering engine that it uses (webkit). We can't afford to ignore mobile users.

Of course we can use simulators and services like browserstack, but these are expensive in terms of both money and time required to test properly.

Cypress allows us to run regular, free tests, and catch bugs fast.

**Comment by paolarosanarodrigues** [comment]
> Some numbers that emphasise the importance of this... in the USA, Safari usage on mobile platforms is something like 52% vs. 40% for Chrome.
> In the UK it's ~48% Safari, ~41% Chrome.
> Where I live in Jersey, Safari is 67%, Chrome 25%.
> [Source: statcounter.com]
> 
> It's obviously super important to be able to test the dominant browser - or at least the javascript/rendering engine that it uses (webkit). We can't afford to ignore mobile users.
> 
> Of course we can use simulators and services like browserstack, but these are expensive in terms of both money and time required to test properly.
> 
> Cypress allows us to run regular, free tests, and catch bugs fast.

I'm living the same reality. We are migrating our applications zone by zone and in this quarter we are starting US zone.
Safari became mandatory for manual and automated tests. We are trying a POC using CodeceptJS or Selenium.

**Comment by laerteneto** [comment]
I am in a middle of a project decision in a big Multinational company, and Cypress not supporting Safari may be a deal breaker for us, since the number of our clients who use Safari is increasing constantly. 

I really like Cypress, I and wish we could use it in this company, since I like it a lot, and I believe Cypress would get a good visibility with another big company using this nice framework.

I really wish they could give us a date for that, so I would be able to show to the company that this integration is due to some date.

**Comment by derekjwilliams** [comment]
I'm in a similar situation to laerteneto, while I'm trying to recommend Cypress, the lack of Safari support on a roadmap is concerning.

**Comment by denis-domanskii** [comment]
Safari became a new IE, more and more conflicts in standard implementation. Safari just can't keep up with the Chrome at all and situation will get worse. So in 1-2 years Safari support became critical. Now about options.

> **A noninclusive list of directions we could take:**
> 
> * Only support Safari during `cypress run`. This does not really live up to our standards of the experience we want to offer with Cypress. It will degrade the true TDE/debugging experience of Cypress when testing in Safari. It is being considered though.
> * Support a 'Webkit' browser. Webkit is the rendering engine that powers Safari. We could offer testing within a Webkit-powered browser. And extend it to be close to the same as testing in Safari.

I vote for the second option:
- You don't need macOS machines, test can be run in Docker on linux
- Headless mode supported
- DevTools protocol is already implemented by Playwright team (but not merged into upstream yet, so some collaboration is needed)
- Should be cheaper than integrate WebDriver and fight with Glass Panes

The only issue - can I trust WebKit? How close it to Safari? According to Playwright folks, it's pretty close. I tend to believe them, but... Several issues with WebKit (false positive in comparison with Safari) can ruin the whole idea.

Despite of risk I still vote for WebKit.

**Comment by cscheffauer** [comment]
+1 for Safari support!

**Comment by evgenikmarkov** [comment]
+1 for Safari support, would be great to have it!

**Comment by nprosenikov** [comment]
+1 for Safari support!

**Comment by linuxkidvoid** [comment]
Thank @jennifer-shehane and Cypress team for adding Safari onto the roadmap! For most of the Safari use case will be Safari on ipad, hope this will be taken into consideration.

**Comment by weyert** [comment]
> Thanks, everyone, for your feedback and comments.
> 
> We’ve got an update from the Cypress side of things. Support for a WebKit browser is on the roadmap for this year.
> 
> We’ve talked to the Safari developer tools team and have done an initial assessment on what it would take to add WebKit support, and we’re beginning an exploratory spike to better understand the timelines and requirements to get this in. We’ll post another update here when we finish that spike.

What are the plans regarding the difference between Webkit browser and Safari used by Apple? It's great to hear that Webkit  support is arriving. If you want I am happy to share my experience with getting a fork of Webkit engine working with Cypress.

**Comment by derekjwilliams** [comment]
This is amazing!  Thanks so much to the Cypress team, you've made testing much more awesome.  Having Safari in the future will be a big plus.

**Comment by janebalmal** [comment]
Is there any chance to get an approximate date (months / years) when Webkit support will be released?
I'd like to have my automated tests to be written with Cypress, but I need to assure my team that Cypress will surely suport Safari browser in future, cause this is required.

**Comment by jennifer-shehane** [comment]
@janebalmal 

>we’re beginning an exploratory spike to better understand the timelines and requirements to get this in. We’ll post another update here when we finish that spike.

**Comment by vitmalina** [comment]
Have you created a branch for webkit support yet? I want to help with the effort.

**Comment by jennifer-shehane** [comment]
There’s a branch/PR beginning the initial work for Webkit support. https://github.com/cypress-io/cypress/pull/15533

**Comment by weyert** [comment]
@jennifer-shehane  https://github.com/cypress-io/cypress/discussions/15681

**Comment by debasisj** [comment]
is there an ETA for this?

**Comment by rhadwyn** [comment]
> is there an ETA for this?

I am also keen on knowing the progress for this as we are in the process of choosing between Cypress and Playwright at the moment.

**Comment by rhadwyn** [comment]
> is there an ETA for this?

I am also keen on knowing the progress of this support as we are in the process of choosing between Cypress and Playwright.

**Comment by rhadronq** [comment]
> is there an ETA for this?

I am also keen on knowing the progress for this support as we are in the process of choosing which tool to use between Cypress and Playwright.

**Comment by onimougwo** [comment]
Is there an ETA for this work ? I don't see any updates since end of march ?

**Comment by metasean** [comment]
Tangential question for individuals who are down-voting the ETA and update requests.

People who are evaluating a tool for potential use, or who are planning features for products that use, a upstream dependency often need to know when to expect something to be in that dependency simply so that they can make plans that are dependent on that dependency.  So I see no reason _not_ to occasionally ask how things are going and when to expect something to be completed.  

It's not a commitment, it's an estimate.  
It's not a demand for a feature.
It's not a contractual deadline, it's just trying to gauge progress.
It's simply asking about the timeline and expectations of something that _the team has already said they're planning to make progress on._  

So what is the reason for down voting requests for an ETA and updates?

**Comment by Floaf** [comment]
People should probably not invest in/evaluate something based on future features/updates since they might never happen. 
Evaluate it for what it is today and if new features arrive in the future, concider it a bonus.

**Comment by 8thblock** [comment]
it comes for fee, mate! Say thank you for their hard free work and be patient

**Comment by metasean** [comment]
@Floaf
First and foremost, *_thank you for your answer!_*

I can't help but be reminded of something I experienced many, many years ago.  At the time I was a poor college student, and iPods had black and white screens, although there had been rumors for quite some time that Apple was going to be releasing a color iPod, and it was a very logical evolution of the iPod.   

I don't ever remember talking to my grandmother about iPods, but my best guess is that I mentioned them to her the last time I saw her.  After her passing, my mother told me that my grandmother had wanted my mother to use whatever was left of her "estate" (I think it was less than $2k, and it was being split between her 2 daughters and 4 grandchildren, so, _not a lot_) to get something for my brother and I that we would really enjoy, and she had made suggestions to my mother about what she thought we might really enjoy.  

My grandmother's suggestion for me was a color iPod.  So after my grandmother's memorial service my mom asked where she could buy a color iPod.  I told her they weren't out yet, so my mother told me to buy myself one when they came out and she would send me the money for it.  When I got back to my campus, I went to the campus store and talked to the guy who handled all the Apple sales.  He told me he had heard the rumors but wouldn't give them much credence. Nevertheless, I decided to wait. Every week or so, I'd go back in, and he and I would have the same conversation.  This went on for a couple of months before I finally broke down and bought a black and white iPod.  Apple literally released the color iPods _the very next day_ —less than 24 hours after I bought the black and white version.  

In other words, even as I was physically purchasing that black and white iPod, the sales guy already had the color iPods behind the counter.  He didn't know that's what was in the boxes, but he absolutely did know that there was a new, unannounced Apple product in those boxes, and he absolutely knew that Apple was going to be making an announcement the next morning about whatever the new product was.  To his credit, after the announcement, when I walked back into the store, quite livid, he very literally pulled a box with a new color ipod from behind the counter because he'd assumed I would be back, and he very happily swapped out the two iPods for me.  So, in the end, I did end up with the color iPod, and every time I used it, I thought about my grandmother's desire for us to be happy.  

I've also thought a lot about how happy I would have been just to look forward to having the color iPod —even while I waited for it to be released— if only I had known _when_ it was going to be released.  Hell, even if it wasn't going to be released for several more years, I would have happily waited.  It was the not knowing that was the most irksome aspect.

**Comment by metasean** [comment]
@8thblock
1st - I am, in fact, extremely grateful for everything the Cypress team does.

2nd - Cypress isn't free for everyone, and since a decision for potential paying customers might be based on this feature, they absolutely have the right to ask.  While I'm not _currently_ a paying customer, but I did help make the decision at my last company to by a subscription.  And I know that this feature is absolutely a blocker for others what might otherwise become paying subscribers.

3rd - Asking for an _estimate_ isn't being impatient.  At least for me, it's quite the opposite.  If I know not to expect something for another 3 to 6 months, then for the next 3 to 6 months I'm excited about the thing.  If in 3 months, there isn't some type of update, then I start to get anxious, but that's still not impatience; however, if I do get an update, particularly one explaining the cause of the delay, my respect goes up for the person/s giving the update.

**Comment by jiangao1** [comment]
Looking at Cypress' roadmap, it's safe to bet Safari support won't happen anytime soon. The Cypress team surely have a lot of other features to develop on their plate given the priority. 

Here are the things to come, looks like Safari support is not something going to happen soon:

https://docs.cypress.io/guides/references/roadmap#Upcoming-features

**Comment by 8thblock** [comment]
WebKit = Safari

**Comment by metasean** [comment]
@jiangao1

fwiw: The last entry in the table of the page that you linked to _IS THIS ISSUE_ and "is in progress". 😉

"I like how people are passionate debating Cypress feature releases. It shows we all care about its success!"

!["gospel truth"](https://external-content.duckduckgo.com/iu/?u=http%3A%2F%2Fawesomelytechie.com%2Fwp-content%2Fuploads%2F2013%2F08%2FTruth-gospel.gif&f=1&nofb=1)

**Comment by flotwig** [comment]
Hi everyone, I wanted to let you all know that this is still on the [roadmap](https://docs.cypress.io/guides/references/roadmap), but it is being blocked by other in-progress work that is higher priority. The work is still assigned to me, so I will return to it as soon as I am unblocked. Rest assured that we are serious about our commitment to add WebKit testing to Cypress.

**Comment by weyert** [comment]
@flotwig Would you accept a PR to add Playwright support to Cypress for selected browsers at this time? E.g. firefox and chrome? I

**Comment by flotwig** [comment]
> @flotwig Would you accept a PR to add Playwright support to Cypress for selected browsers at this time? E.g. firefox and chrome?

@weyert I have not done too much research into driving the Playwright FF/Chrome, only the Playwright WebKit. I am interested in this idea though, Playwright's model of distributing patched browser binaries is something we have discussed adopting for Cypress as well, since it removes the blocker of having to wait on the upstream browser vendors to make their browsers "testable". Potentially this is the direction Cypress will move in the future, away from having users supply their own browsers.

Could you open up a new issue, describing what you had in mind, and link it to me here? I would love to discuss this with you more.

**Comment by weyert** [comment]
I have raised ticket see: https://github.com/cypress-io/cypress/issues/16996

**Comment by weyert** [comment]
Everyone, I have made a work in progress PR which allows you to run your Cypress tests using the Playwright browsers (firefox and chrome atm). It will also list webkit int he browsers list but its not really functioning

**Comment by nghedia** [comment]
@weyert  great work on adding playwright support.. I have managed to get webkit working based on your PR. The issue is certainly the socket.io connection.. I resolved the 400 error by changing the 'transports' property from websocket to polling.

<img width="1920" alt="Screenshot 2021-06-25 at 01 15 45" src="https://user-images.githubusercontent.com/20408351/123351026-2c2c4500-d554-11eb-8a9c-f9b788101c05.png">

**Comment by weyert** [comment]
@nghedia Glad to hear you got it working. I had this on my to do list but work got in the way.

 I did notice that the 400 is probably cause because Webkit is trying to connect to the web socket while by passing the proxy. This appears a bug in Playwright in some cases but its unclear if the case applies to  Cypress. Maybe @flotwig or Cypress can shed some light on that.

For now, I will apply the same change to my PR and make it work. My assumption this is not the wanted approach by Cypress :) I have been working to get downloads working too.

Looking forward to hear from you, if your tests are working as expected

**Comment by weyert** [comment]
@nghedia Yeah, I have tried it with `polling`-transport that's definitely working. I hope I can somehow assist playwright with fixing the proxy issue. I will try if with websockets works when using the system proxy 🤔

**Comment by melibe23** [comment]
I'm gonna add +1 to people requiring Safari support in Cypress. 20% of our users use this browser (most of them are from US)

**Comment by jdferreira2429** [comment]
Same here, +1 on needing safari support.

**Comment by weyert** [comment]
@melibe23 @jdferreira2429 Feel free to try out my PR

**Comment by hogarthww-labs** [comment]
@weyert Any movement or updates on the Playwright proxy issue?

**Comment by cory4patriots** [comment]
**Same here, +1 on needing safari support.**

**Comment by weyert** [comment]
@Aarbel Let's first get Webkit working on Desktop before worrying  about adding more complexity by trying to support mobile. You probably can get away with emulating device metrics, geolocation, orientation 

I can imagine a limited feature set of Cypress could be possible to get working on MobileSafari. I am curious why would you be interested in MobileSafari testing? Also is WKWebView suffice for you? Planning to test those navigation bar weirdness? Also would you want to test on a real device or a the iOS simulator?

**Comment by weyert** [comment]
> @weyert Any movement or updates on the Playwright proxy issue?

Yeah, I am currently blocked by how to add an experimental flag in the code base to allow opting into using Webkit. For now, I decided to just use the fallback non-websocket strategy for `socket.io`. But after current code is behind the flag I want to revisit it.

**Comment by weyert** [comment]
Sure, it basically means that it's difficult to use Apple Safari in a way to make it work with Cypress. Instead the underlying engine used by Safari named Webkit is used. Normally, it's safe to use Webkit engine build to target Safari as this would also allow you to run Cypress tests in a Safari-like browser in Windows, Linux.

**Comment by VictoriaBello** [comment]
+1 on safari support, my company just decided to go with another framework just for this. Quote: "back to the main question which framework we should choose for automation, so from my understanding if the framework is not capable of running automation on Safari is a deal-breaker, even though the issue was reported with manual QA was always a minor issue on safari.  so  cypress framework is not a good candidate."

**Comment by VictoriaBello** [comment]
> To clarify it will be WebKit support not Safari support and only Desktop. I have been using it for months Cypress+WebKit for months now without major issues for Cypress functionality I use

To my understanding... using webKit, can only validate that your app loads with Safari, but doest not run all tests there. 
Got that info from: https://medium.com/tech-learn-share/shall-we-integrate-cypress-io-with-safari-browser-c190d600e771

Am I wrong? That page might not be updated....

**Comment by romankhomitskyi** [comment]
> > To clarify it will be WebKit support not Safari support and only Desktop. I have been using it for months Cypress+WebKit for months now without major issues for Cypress functionality I use
> 
> To my understanding... using webKit, can only validate that your app loads with Safari, but doest not run all tests there. Got that info from: https://medium.com/tech-learn-share/shall-we-integrate-cypress-io-with-safari-browser-c190d600e771
> 
> Am I wrong? That page might not be updated....

The browsers provided by Playwright is the best of what you can get for e2e testing. Running WebKit on a Linux CI would use a different compositing pipeline than macOS, but it can be used for e2e testing of web applications in WebKit-powered browsers.

**Comment by weyert** [comment]
@romankhomitskyi Yeah, I was able to get most of it working. I still waiting for assistance by @flotwig or Cypress developers to put things behind a feature flag as was requested earlier. I just noticed that @BlueWinds said is working on it too.

**Comment by vladlearns** [comment]
To all of you who are waiting for https://github.com/cypress-io/cypress/pull/17002, here is a small workaround: https://github.com/vladlearns/cypress-playwright. This will let you execute playwright from cypress test. @mdusafp, maybe this will help. Cheers!

**Comment by romankhomitskyi** [comment]
> I am still waiting for assistance from Cypress. I don't understand why you don't want to help me

it's simple, they do not wanna integrate resourses from other competitors))

**Comment by kaushikmehta** [comment]
Requesting support for Safari as well

I am trying to convince my leadership that this is the best tool and will soon be an industry benchmark (if not already) but constantly get asked if I am sure because the perception is that this tool is 'incomplete' as it does not support Safari - obviously I don't think that is the case but I am afraid I am going to get asked to drop my POC in favour or Playwright, TestCafe or Selenium WebDriver (none of which I prefer)

To the Cypress team, much ❤️  for what you have built so far and really hoping that you can empower us to execute our tests on Safari soon!

**Comment by jiangao1** [comment]
Totally agree with @kaushikmehta . Not supporting safari is quickly becoming a marketing problem. Unfortunately many leadership, none-technical managers and bill payers alike can only grasp the concept of browser compatibilities during tool selection discussions. They will use the only concept they understand to derail engineer choices. I've seen many and heard many such cases. Would be nice if Cypress can allocate some marketing budget over to add more engineering work in this area. Just a suggestion. I am a happy Cypress user, and my team doesn't care about Safari. I'm only speaking for others who feel the pain right now.

**Comment by vladlearns** [comment]
Same here. I really like the tool, but given that there is no support for WebKit, I need to do a mix of things or go with something else. I see how adding the WebKit support can literally change the whole game.

**Comment by Sam-Lin-MillersLab** [comment]
This is a classic case where engineers don't understand marketing. The whole point of E2E testing is for e-comm and like it or not, Apple users are ones with deep pocket and willing to spend money. (Very obvious right? they can spend so much on a phone already.) So like it or not, when we programmers are trying to sell Cypress to our shareholders, they will definitely ask does it support Apple users. So I am really really surprised this is not a top priority.

**Comment by romankhomitskyi** [comment]
> This is a classic case where engineers don't understand marketing. The whole point of E2E testing is for e-comm and like it or not, Apple users are ones with deep pocket and willing to spend money. (Very obvious right? they can spend so much on a phone already.) So like it or not, when we programmers are trying to sell Cypress to our shareholders, they will definitely ask does it support Apple users. So I am really really surprised this is not a top priority.


It's not top priority because safari API is difficult and different, even Playwright built their own browser on top of Webkit

**Comment by dantman** [comment]
It's a problem for those of us implementing things too trying to ensure browser specific bugs don't leak through. Though Safari alone wouldn't solve that, the mobile browsers sometimes also have different behaviour to their desktop counterparts.

My client and I were only convinced to use Cypress because we didn't want to run all the tests on all browsers anyways. So I decided to write all the tests that test the app itself in Cypress and re-write the subset of tests that were also browser compatibility smoke tests as duplicate TestCafe tests.

This all of course would be less of a pain point if everyone used a common test API. Instead of requiring you to commit to a framework's API up front and leave you to rewrite tests entirely if you need to switch. If we had a common API then I could just develop tests in Cypress, run those Cypress for the PR CI, and run the subset of browser compatibility smoke tests in TestCafe.

Of course, as it stands now, if TestCafe ever managed to clean up their DX so it was on par with Cypress, Cypress would probably get tossed out the window. The tiered methodology where you can do with directly integrated browsers but can still run those same tests with a subset of the functionality on less integrated browsers is great.

**Comment by Sam-Lin-MillersLab** [comment]
@romankhomitskyi 
But they did it at a certain level. Right? And they made it very clear as a goal from the beginning. I don't know if we should put difficult & different as a problem here. If it is easy, it is not worth doing. 😄 

I hate to say this but if Cypress not being bundled by NX and Storybook, I won't even bother to say a word here. And I have tried to push NX to support Playwright. I really hope Cypress can make this work.

**Comment by weyert** [comment]
I can only say it's possible to have Cypress+Webkit. Cypress has decided to go a different route than the proposal I made through my PR.  It's a big bummer for me but that's life.

**Comment by romankhomitskyi** [comment]
> I can only say it's possible to have Cypress+Webkit. Cypress has decided to go a different route than the proposal I made through my PR. It's a big bummer for me but that's life.

You did a great job

**Comment by dsmileybrainly** [comment]
Is this still not under active development?

**Comment by jiangao1** [comment]
> Is this still not under active development?

It has been under consideration for a few years. At one time, they were spiking the effort for a few months, but then went silent. My guess is the company doesn't think it's worth the effort the support safari.

At this point, given the track record, I won't count on this for a few years. There are other tools innovate fast as well in the javascript land such as TestCafe and Microsoft PlayWright. I played around with TestCafe for Safari testing, it worked great.

**Comment by dsmileybrainly** [comment]
> > Is this still not under active development?
> 
> It has been under consideration for a few years. At one time, they were spiking the effort for a few months, but then went silent. My guess is the company doesn't think it's worth the effort to support safari.
> 
> At this point, given the track record, I won't count on the support of Safari for a few years. There are other tools innovate fast as well in the javascript land such as TestCafe and Microsoft PlayWright. I played around with TestCafe for Safari testing, it worked great.

They do have it listed on the current roadmap though. Last updated in Sept 2021. If anyone from the Cypress team out there can comment with an update that would be very helpful.
<img width="712" alt="Screen Shot 2022-03-28 at 7 17 05 AM" src="https://user-images.githubusercontent.com/69155578/160422152-dab64a5a-b180-49ab-9d3b-b2fb678d9047.png">

**Comment by pdrprz** [comment]
@flotwig thanks for the update! Is there a date for when Cypress 10.0 will be launched?

**Comment by BryantUng-at** [comment]
Just wanted to check back again if there was a timeline for this support?

**Comment by lughino** [comment]
> I have given up the hope and switching to playwright.

Same here. After long fights to keep Cypress in the company, I have given up.
The Playwright's maturity is now very close to Cypress in terms of developer experience.
No way I'll come back to Cypress.

**Comment by hitmands** [comment]
I'd switch to playwright right now if I'd find a good alternative to the dashboard.

**Comment by acourtiol** [comment]
Same here, canceled our Cypress dashboard subscription, moved everything to Playwright, not regretting the move after seeing Cypress 10 announcement.

**Comment by kfsass** [comment]
Yup, another voice here whose company is abandoning Cypress in favor of Playwright for WebKit support. Figure out how to get your developers working on other initiatives working on this one. It would be a shame to see Cypress left behind but it boggles the mind how consistently browser support beyond Chrome has been deprioritized by this team.

**Comment by weyert** [comment]
> This is something should be taken on priority basis, people from QA background moving to playwright as cypress doesn't support Safari, Being an end user of cypress, I really want cypress to make it possible so that me and my team don't need to switch to playwright or any other tool as I really like some of amazing features that cypress has, hoping and looking forward to have safari support in near future, good luck. @flotwig

The Playwright community is great, the maintainers are quick with responding to issues. You should consider switching to Playwright. Especially with the recently introduced component testing support. What kind of other amazing features are you missing in Playwright?

**Comment by kaustubhshete** [comment]
For folks who are looking for dashboarding capability,  you can use playwright with reportportal.io. Reportportal has custom integration with playwright.
All the reporting capability plus playwright's parallel execution capability provides a killer alternative to Cypress and both are open source offerings.

**Comment by jamesopti** [comment]
> @dsmileybrainly @jiangao1 WebKit is still planned for development. We have a clear plan of where we want to go with it and how to build it, but currently the developers assigned to WebKit are working on other initiatives. We know how important this is to users and it's one of the first things we'll be focusing on after the upcoming Cypress 10.0 launch.

@flotwig Now that Cypress 10 has been released and is stable, are there plans to implement this? Any updates on progress?

**Comment by weyert** [comment]
While I still maintain a private fork of Cypress with WebKit support but I would suggest to switch to a different solution like Playwright. I am actively migrating my Cypress tests to Playwright and so far it’s a breeze to work with. Already converted them all at work but some passive income projects still run Cypress

**Comment by dsmileybrainly** [comment]
@jamesopti is there any movement on this issue? We've made significant investments at Brainly to support Cypress due to its integration in the NX ecosystem and the promise of Safari support. We will have to migrate away from Cypress (at a heavy cost) if this is not delivered as promised.

**Comment by jamesopti** [comment]
> @jamesopti is there any movement on this issue?

Not sure if you intended to ping me as the person to ask, but I am a consumer of Cypress and am wondering the same thing myself.

**Comment by dsmileybrainly** [comment]
> @dsmileybrainly @jiangao1 WebKit is still planned for development. We have a clear plan of where we want to go with it and how to build it, but currently the developers assigned to WebKit are working on other initiatives. We know how important this is to users and it's one of the first things we'll be focusing on after the upcoming Cypress 10.0 launch.

Is there any movement on this issue? We've made significant investments at Brainly to support Cypress due to its integration in the NX ecosystem and the promise of Safari support. We will have to migrate away from Cypress (at a heavy cost) if this is not delivered as promised.

**Comment by jamesopti** [comment]
@jennifer-shehane Would be great to hear _something_ from the core Cypress team on this one. Seems like a pretty popular  issue (188 comments) that the community is eagerly waiting for (even if the truth is that it is not yet being developed yet).

**Comment by BlueWinds** [comment]
Jennifer is on leave right now, will be back later this year, but I can provide an update: we weren't actively working on this before, but we are now. @flotwig had an almost-complete PR that was put on hold several months ago, but since the release of 10.x, it's fallen pretty far out of date and he's had to start over.

He can speak to the complexity involved, but it *is* being actively worked on and we should expect some progress (a PR ready for review) fairly soon here.

**Comment by weyert** [comment]
Really curious how Cypress solves the problem, especially after you decided to avoid using Playwright WebKit and wanted to manage your own WebKit app build for all your supported platforms. Are you planning to open-source that too?

**Comment by flotwig** [comment]
> Really curious how Cypress solves the problem, especially after you decided to avoid using Playwright WebKit and wanted to manage your own WebKit app build for all your supported platforms.

@weyert This will use `playwright-webkit`. Yes, it's the same approach as before, that you and I were both trying in our PRs. Like you're referencing, we intended to block WebKit support until we had the infrastructure to build+distribute our own patched browsers ready. Internal priorities shifted and moved that goal into the future, and WebKit support with it, hence the lack of updates here.

> Are you planning to open-source that too?

That's the plan.

---

I'm sorry, everyone, that it has taken this long to get this feature back on the top of Cypress's priority list. But I am also happy knowing that we will be able to unblock y'all soon with `experimentalWebKitSupport`. Stay tuned for updates here.

**Comment by billyromano** [comment]
Knowing that Cypress will be supporting WebKit this year (hopefully sooner than later) will be the leverage to help my company decide with how to handle E2E testing for our clients👍😎

Microsoft, Google and Apple should all be supported for E2E testing since Cypress is the only premium and enterprise offering in my research. We have a high use case for all supported browsers and depending on some clients/projects Apple is used more than Google and Microsoft combined.

**Comment by dsmileybrainly** [comment]
> Knowing that Cypress will be supporting WebKit this year (hopefully sooner than later) will be the leverage to help my company decide with how to handle E2E testing for our clients👍😎
> 
> Microsoft, Google and Apple should all be supported for E2E testing since Cypress is the only premium and enterprise offering in my research. We have a high use case for all supported browsers and depending on some clients/projects Apple is used more than Google and Microsoft combined.

I could not agree more! I am in this same situation. Knowing support is coming hopefully this year gives me leverage to stay committed to the only enterprise offering.

**Comment by nancyluvend** [comment]
Thanks so much! Keen to use it!

**Comment by kopepasah** [comment]
First off, thanks for the Cypress team for finally jumping on this, as they can see it IS prudent to the future success of this project to support testing in Safari.

After reading through this thread, I personally get the feeling that the Cypress team was very reluctant to implement this feature, even going as far to think we want this feature to appease C-suite executive requests. However, this cannot be further from the truth, as Safari usage has been growing consistently year over year, and is now the second most used browser on the web. 

We are asking for this feature because we need to run tests in browsers our users use, not our developers. While the challenges of testing in Safari (and Webkit) are present, the thought of not testing in these scenarios because the tooling is hard is a no-go for many companies (there's a reason Microsoft uses Playwright.dev).

As I stated before, I am excited to see support for Safari coming, but I do hope the Cypress team learns from this experience and prioritizes testing where users are, instead of what tools developers use.

**Comment by weyert** [comment]
> After reading through this thread, I personally get the feeling that the Cypress team was very reluctant to implement this feature, even going as far to think we want this feature to appease C-suite executive requests. However, this cannot be further from the truth, as Safari usage has been growing consistently year over year, and is now the second most used browser on the web.

> As I stated before, I am excited to see support for Safari coming, but I do hope the Cypress team learns from this experience and prioritizes testing where users are, instead of what tools developers use.

My experience is that adding WebKit support to Cypress wasn’t straightforward at all. Of course, I never knew the code base as well as Cypress employees but this browser behaves in some ways differently from Firefox and Chrome in my experience so you end up with special logic for WebKit. For example, something as the type() action in Cypress. So I can understand reluctance, but I am glad it’s coming.

**Comment by romankhomitskyi** [comment]
> > After reading through this thread, I personally get the feeling that the Cypress team was very reluctant to implement this feature, even going as far to think we want this feature to appease C-suite executive requests. However, this cannot be further from the truth, as Safari usage has been growing consistently year over year, and is now the second most used browser on the web.
> 
> > As I stated before, I am excited to see support for Safari coming, but I do hope the Cypress team learns from this experience and prioritizes testing where users are, instead of what tools developers use.
> 
> My experience is that adding WebKit support to Cypress wasn’t straightforward at all. Of course, I never knew the code base as well as Cypress employees but this browser behaves in some ways differently from Firefox and Chrome in my experience so you end up with special logic for WebKit. For example, something as the type() action in Cypress. So I can understand reluctance, but I am glad it’s coming.

I guess iframes in webkit it's a long shot

**Comment by flotwig** [comment]
> If I remember correctly, you somewhere needed to get the process id, to properly integrate the webkit browser in the Cypress system. My own code of Cypress suggests that the following should help getting access to that:

@weyert Oh yeah, this is way better than how I was thinking of doing it, thanks. I was going to pull the full Cy process tree and try to identify the PW process using the command string. We need the pid for `process_profiler`. But there's a 2nd use for it. I noticed that `await browser.close()` can take several seconds to complete, so we might want to use `process.kill(webkitPid)`  to end the WK process instead. So this is doubly useful.

> In my version I never really cleanly got the video recording integrated

Currently working on this, and it's taking somewhat of a refactor of how Cypress records video. Cypress expects to be able to start video recording anytime and to receive raw frame data, and PW does neither of those :stuck_out_tongue:

> I will see if I make some PRs of working bits that I have that you still need to implement after I met my work deadline (e.g. downloads)

Great, if you do have time, the task list on this PR should be current. Feel free to ping me on Discord if you'd like to coordinate efforts.

**Comment by monkpit** [comment]
Closed means it's going to ship soon? 👀 we have had our eye on this issue in order to expand our Cypress tests to encompass Responsive apps on Safari. I'd be overjoyed if this is going to ship :) Thanks for all of your work!

**Comment by weyert** [comment]
Great  job on closing this ticket. Looking forward seeing it announced :)

**Comment by richard-owusu** [comment]
Good job! When can we expect the announcement?

**Comment by flotwig** [comment]
Hey folks, this is coming out as experimental in today's release, 10.8.0. I'll comment back here once it's out.

**Comment by flotwig** [comment]
Cypress 10.8.0 is now available with `experimentalWebKitSupport`!

See the docs for instructions on installing and enabling WebKit: https://docs.cypress.io/guides/guides/launching-browsers#WebKit-Experimental

There are some known issues under the [experiment: webkit](https://github.com/cypress-io/cypress/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3A%22experiment%3A+webkit%22) tag, please consult those before opening a new issue.

https://www.cypress.io/blog/2022/09/13/cypress-10-8-experimental-run-tests-in-webkit/
